client/components/tinymce/plugins/media/plugin.jsx

```diff
@@ -190,22 +276,50 @@ function mediaButton( editor ) {
 					return;
 				}
 
+				let useMediaSize = false;
+
+				if ( media.isDirty ) {
+					// If an image is edited through image editor and its final size is smaller than the size of
+					// the inserted image, let's update the size of inserted image to the size of edited image.
+					useMediaSize =
+						media.width < current.media.width ||
+						media.height < current.media.height;
+				}
+
 				// When merging, allow any updated field to be used if it doesn't
 				// already exist in the current markup, but otherwise only force
-				// update ID and URL attributes to their new values
-				const merged = assign( {}, media, current.media, pick( media, 'ID', 'URL' ), {
-					transient: !! media.transient
-				} );
+				// update ID, URL, width and height attributes to their new values
+				const merged = assign(
+					{},
+					media,
+					current.media,
+					pick(
+						media,
+						'ID',
+						'URL',
+						useMediaSize ? 'width' : '',
+						useMediaSize ? 'height' : ''
+					),
+					{
+						transient: !! media.transient
+					}
+				);
 				const options = assign( {}, current.appearance, {
 					forceResize: ! media.transient && current.media.width && current.media.width !== media.width
 				} );
 
-				// Since we're explicitly targetting the image, don't allow the
-				// caption to be considered in the generated markup
-				delete merged.caption;
+				if ( ! mediaHasCaption ) {
+					// Since we're explicitly targetting the image, don't allow the
+					// caption to be considered in the generated markup
+					delete merged.caption;
+				}
 
 				// Use markup utility to generate replacement element
 				markup = MediaMarkup.get( merged, options );
```

Comment:
If `merged` contains `caption` prop, the `img` tag will get wrapped in the caption shortcode with correct sizing.